### PR TITLE
Eliminate some allocations in client_channel code.

### PR DESCRIPTION
**Pull request by markdroth**
This is the part of #11298 that isn't blocked on combiner improvements.

### CODE REVIEW

**Comment by markdroth** [comment]
This also includes some code cleanups.

**Comment by dgquintas** [comment]
I'll review this soon. I'd however like it if #11159 goes in first in order to prevent conflicts in an already very large PR.

**Comment by markdroth** [comment]
No problem -- I'm fine waiting on the other PR.

File: src/core/ext/filters/client_channel/client_channel.c [hunk 1]
```diff
@@ -774,6 +735,8 @@ static void cc_destroy_channel_elem(grpc_exec_ctx *exec_ctx,
 
 #define CANCELLED_CALL ((grpc_subchannel_call *)1)
 
+#define MAX_WAITING_BATCHES 6
```
Comment (dgquintas) [inline src/core/ext/filters/client_channel/client_channel.c]: where's that magic 6 coming from? Why 6 and not 10? We should at the very least have a comment. Ideally we'd get that max number from some other file where it's uniquely defined.

File: src/core/ext/filters/client_channel/resolver/sockaddr/sockaddr_resolver.c [hunk 2]
```diff
@@ -88,7 +88,9 @@ static void sockaddr_shutdown_locked(grpc_exec_ctx *exec_ctx,
   sockaddr_resolver *r = (sockaddr_resolver *)resolver;
   if (r->next_completion != NULL) {
     *r->target_result = NULL;
-    grpc_closure_sched(exec_ctx, r->next_completion, GRPC_ERROR_NONE);
+    grpc_closure_sched(
+        exec_ctx, r->next_completion,
+        GRPC_ERROR_CREATE_FROM_STATIC_STRING("Resolver Shutdown"));
```
Comment (dgquintas) [inline src/core/ext/filters/client_channel/resolver/sockaddr/sockaddr_resolver.c]: Could you also replicate this change in fake_resolver.c while at it?

File: src/core/ext/filters/client_channel/client_channel.c (hunk 1 above)
Comment (markdroth) [inline src/core/ext/filters/client_channel/client_channel.c]: Done.

File: src/core/ext/filters/client_channel/resolver/sockaddr/sockaddr_resolver.c (hunk 2 above)
Comment (markdroth) [inline src/core/ext/filters/client_channel/resolver/sockaddr/sockaddr_resolver.c]: Done.

**Comment by markdroth** [comment]
I had actually already merged in the LB update change, but there was another one that was merged into master right before I pushed the changes to this PR.  I'll update again and see what the tests say.

**Comment by markdroth** [comment]
Looks like there's an assertion failure triggering in the Windows Node tests.  I don't understand how it's happening, though, since I can't find a code-path that would cause that.

@murgatroid99, is there a way for me to run those tests such that the output includes the C-core logs?  If so, I can add some debug logging to try to figure out how this is happening.

**Comment by markdroth** [comment]
This PR should be in good shape now.  David, PTAL.

**Comment by markdroth** [comment]
Known issues: #11055 #11223

**Pull request merged**